Restart option added in case of unexpected termination

sakim8048 opened the pull request:
This PR includes two major updates. Updated restart option will restart optimization or vibration calculations if previous runs are unexpectedly terminated. Also this PR will allow users to run Pynta efficiently on ALCF Polaris machine. Details are described below:

---------------------------------------------------------------------------------------------------------------------------- 

1. Restart Option

Upon running `restart(),` it will retrieve Fireworks workflow information, including the workflow ID number, task ID number, task states, and launch directories where unexpectedly terminated calculations were running. Before rerunning Fireworks for the incomplete runs, all necessary files, such as optimization trajectory files or vib folders, will be copied and sent to the destination directory.

If task states are not completed (e.g., fizzled or lost runs), the optimization runs will restart from the last geometry of the optimization trajectory file in the previous launch directory. In the case of a vibration restart, empty vibration JSON files will be deleted from the vib folder before rerunning the vibration.

2. Running Pynta on ALCF Polaris

With Raymundo's efforts, this PR allows Pynta to run on ALCF Polaris with a single queue allocation. Raymundo updated the way Pynta maps tasks on each node for ALCF machines. Each task runs on a different Fireworker, and each Fireworker is associated with a node. This is available for `multilauncher`. The optimal approach is to set `num_jobs` in Pynta input script to the number of nodes.

tdprice-858 committed:
Function for generating minimum energy adsorbate db with all data required for thermochemistry

tdprice-858 committed:
Made base thermochemistry class for Pynta

tdprice-858 committed:
Adding method to thermochemistry class to calculate heats of formation referenced to gas phase adsorbates. Current code is intended for gas phase refs: H2O, CH4, H2, NH3

tdprice-858 committed:
adding method to thermo class calculate translational thermo

tdprice-858 committed:
Added method to thermo class for calculating vibrational thermo

tdprice-858 committed:
Add method to thermoclass to fit NASA polynomial from thermochemistry data

tdprice-858 committed:
Add method to thermo class for formatting RMG library output

tdprice-858 committed:
Added method to thermochem class for generating RMG thermochem library header

tdprice-858 committed:
Added method to thermo class for calculating temperature corrections, calling subroutines for fitting NASA polynomials and formatting RMG output.

mjohnson541 committed:
correctly find index of the site in adslabmol to provide to add_coadsorbate_2D

mjohnson541 committed:
fix naming issues in TSest task adaption

mjohnson541 committed:
add safety ensuring Molecule updating in geometricanalysis

mjohnson541 committed:
update Molecule objects after merge actions

mjohnson541 committed:
use copies of the 2D adslabmol when looping through possible sites

rayhe88 committed:
polaris mapping info added in polaris.py

rayhe88 committed:
launch several Rockets in a parallel environment

rayhe88 committed:
Get information of the workflow and restart Pynta for optimization restart job. Delete empty vibration cache file in vib folder for vibration restart job

rayhe88 committed:
Get information of the workflow and restart Pynta for optimization restart job. Delete empty vibration cache file in vib folder for vibration restart job

sakim8048 committed:
add machine keywords to optimization and vibration firework tasks

add machine keyword in firework task

Function copyDataAndSave() is added to copy a file from a origin subdirectory to a destination subdirectory

add machine keyword to vibration keywords in setup_adsorbates

sakim8048 committed:
add machine keyword in firework task

sakim8048 committed:
add machine keyword in firework task

sakim8048 committed:
add machine keywords to optimization and vibration firework tasks

add machine keyword in firework task

Function copyDataAndSave() is added to copy a file from a origin subdirectory to a destination subdirectory

add machine keyword to vibration keywords in setup_adsorbates

rayhe88 committed:
add machine keywords to optimization and vibration firework tasks

Function copyDataAndSave() is added to copy a file from a origin subdirectory to a destination subdirectory

add machine keywords to optimization and vibration firework tasks

add machine keyword in firework task

Function copyDataAndSave() is added to copy a file from a origin subdirectory to a destination subdirectory

add machine keyword to vibration keywords in setup_adsorbates

add machine keywords to optimization and vibration firework tasks

rayhe88 committed:
Function copyDataAndSave() is added to copy a file from a origin subdirectory to a destination subdirectory

rayhe88 committed:
Function copyDataAndSave() is added to copy a file from a origin subdirectory to a destination subdirectory

mjohnson541 committed:
Calculate IRC with fmaxirc not fmaxopt

It seems that https://github.com/zadorlab/pynta/pull/45 somehow changed the fmax used in the IRC calculation from the intended fmaxirc to fmaxopt.

sakim8048 committed:
restart module update: copy traj file to destination directory for opt task and delete empty json file in vib folder for vib task

mjohnson541 committed:
specify flask in environment.yml

The fireworks gui will not display workflows with newer versions of flask.

mjohnson541 committed:
add fischer-tropsch run for testing

mjohnson541 committed:
update split_adsorbed_structures
1) ensure bonds between surface sites are removed
2) add the split_sites_with_multiple_adsorbates option to split sites in cases where where multiple adsorbates are bonded to the same site

mjohnson541 committed:
add a reaction_adsorbate_bond_cutoff to generate_TS_2D

this cutoff prevents generate_TS_2D from creating covalent reaction bonds between atoms that are unreasonably far apart

mjohnson541 committed:
add function validate_TS_geometry

this function assesses the geometry and the imaginary frequency of the TS to determine validity (based on that information)

mjohnson541 committed:
add function validate_TS_ircs to assess TS validity based on IRC information

mjohnson541 committed:
add function validate_TS to combine geometry/frequency information and IRC information when available to overall assess the validity of a TS

mjohnson541 committed:
add validate_TS_configs function
applies TS validation to all individual TS guess runs for a given overall target TS

mjohnson541 committed:
add testGeometricanalysis.py file for testing geometricanalysis.py functions

adds testing file for geometricanalysis.py and tests TS validity functions by running them on the fischer-tropsch set in the testing data

mjohnson541 committed:
add automatic transition state validation information in the postprocessing notebook

mjohnson541 committed:
enable use of any ASE calculator for harmonically forced calculations within pynta

mjohnson541 committed:
take harmonically forced software inputs

mjohnson541 committed:
adapt adsorbate guess generation to software flexibility

mjohnson541 committed:
adapt TSest to harmonically forced software flexibility

mjohnson541 committed:
rename functions to remove xtb specific names

mjohnson541 committed:
remove unnecessary import of HarmonicallyForcedXTB

mjohnson541 committed:
handle TBLite calculator import in name_to_ase_software

mjohnson541 committed:
handle XTB and TBLite the same for testing purposes

mjohnson541 committed:
add test for harmonically forced optimization

mjohnson541 committed:
bugfixes for adsorbate guess generation, particularly for bidentates

mjohnson541 commented on pynta/main.py:
```diff
@@ -586,6 +605,92 @@ def execute_from_initial_ad_guesses(self):
+        self.launch()
+
+#restart option: RHE + KSA
+    def restart(self,wfid):
```
can you add a docstring?

mjohnson541 commented on pynta/main.py:
```diff
@@ -685,7 +685,7 @@ def restart(self,wfid):
                             print(f'No directory named "vib" found in {src}.')
                             print('No vibration calculations executed: Check optimization runs are finished and optimized geometries are collected.')              
-                            
+
```
Remove these whitespace changes.

mjohnson541 reviewed (changes requested):
Thanks! I went over most of it.

mjohnson541 committed:
enable clean_pynta_path to handle not saving the initial guess

mjohnson541 committed:
rename Ntsmin to Nharmmin, remove fmaxopthard

mjohnson541 committed:
add construct_initial_guess_files function

mjohnson541 committed:
enable MoleculeCollect fireworks to either add new fireworks as detours or additions

mjohnson541 committed:
create MolecularAdsorbateEstimate Firetask

generates initial guesses for the input adsorbate
generates fireworks for optimizations and a collect that will trigger the vibration calculations

mjohnson541 committed:
adapt initial guess generation to occur through fireworks

setup_adsorbates now sets up MolecularAdsorbateEstimate task fireworks
atom map generation now occurs separately in the new generate_atom_maps function

mjohnson541 committed:
adapt TSest to change in adsorbate_fw_dict

mjohnson541 committed:
adapt execute to the new workflow

mostly initial guess generation related things are removed

mjohnson541 committed:
remove initial guesses from the main test example

now these initial guesses are generated as a part of the workflow so they don't need supplied beforehand

mjohnson541 committed:
adapt testMain.py to initial guess changes

mjohnson541 committed:
adapt map_harmonically_forced and move to calculator.py

enable it to be used for both harmonic optimizations (adsorbate guess and TS guess)
enable it to be run in parallel with joblib

mjohnson541 committed:
add and pass along the nprocs_harm attribute

this indicates the number of processes that should be used for running harmonic optimizations in parallel

mjohnson541 committed:
parallelize adsorbate guess generation

mjohnson541 committed:
parallelize HFSP optimizations

mjohnson541 committed:
run the test example using 4 processes for harmonic optimizations

tdprice-858 committed:
Update postprocessing.py

removing type hint for ase db which was causing test to fail

mjohnson541 committed:
handle TooManyElectronsException errors when generating unstable pairs

mjohnson541 committed:
fix importing of place_adsorbate

mjohnson541 committed:
restore adsorbate placement handling for covdep

mjohnson541 committed:
improve adsorbate guess generation logging

mjohnson541 committed:
delete forces to avoid errors when constructing Atoms object

mjohnson541 committed:
update the slab used for the main test

the original slab did not have the energy attribute, which caused postprocessing to crash

mjohnson541 committed:
delete generated rmg libraries when cleaning pynta directory

mjohnson541 committed:
add to_dict function

enables conversion of objects to dictionaries

mjohnson541 committed:
remove admol construction error logging for handled errors

mjohnson541 committed:
enable get_vibdata to handle gas phase species

mjohnson541 committed:
improve TS validation information

before whether the TS frequency information was validated was not (directly) in the information dictionary

mjohnson541 committed:
fix get_cp function

it should be either dA/dT in the harmonic case or dH/dT at constant P in the ideal gas case

mjohnson541 committed:
add GasConfiguration for handling statmech and energy calculations for Gas phase species

mjohnson541 committed:
create SurfaceConfiguration object

mjohnson541 committed:
tweak computation of heat of formation

remove prints
assign heat of formation in J/mol

mjohnson541 committed:
tweak translation thermo calculations for SurfaceConfigurations

in particular calculate Cp0_trans and Cpinf_trans

mjohnson541 committed:
tweaks to vibrational thermo calculations

in particular enable calculation of Cv0_vib and Cvinf_vib

mjohnson541 committed:
use rmgmolecule to fit thermochemistry greatly simplifying fitting NASA polynomials

mjohnson541 committed:
improve RMG output/thermo commenting

mjohnson541 committed:
adapt thermochemistry calculations in SurfaceConfiguration

mjohnson541 committed:
add run function for running statmech calculations for SurfaceConfiguration

mjohnson541 committed:
add new Class Kinetics for running kinetics calculations

mjohnson541 committed:
add to_dict and calculate_barrier for Kinetics

mjohnson541 committed:
enable calculation of kinetic parameters in Kinetics objects

mjohnson541 committed:
enable generation of RMG reaction library entries

mjohnson541 committed:
add run function for Kinetics

mjohnson541 committed:
add functions for calculating reaction thermochemistry

mjohnson541 committed:
add get_species function for postprocessing species into configuration dictionaries

mjohnson541 committed:
add get_TS function for postprocessing TSs into configuration dictionaries

mjohnson541 committed:
add function get_kinetics for postprocessing TSs into Kinetics objects

mjohnson541 committed:
add get_reference_energies function for extracting reference energies from pynta run

mjohnson541 committed:
add postprocess function for postprocessing Pynta runs into Configuration and Kinetics objects

mjohnson541 committed:
add write_rmg_libraries function for writing RMG libraries for a pynta run

mjohnson541 committed:
add Postprocessing Firework for running postprocessing automatically during Pynta runs

mjohnson541 committed:
enable Collect fireworks to generate Postprocessing fireworks

mjohnson541 committed:
add postprocessing options to the adsorbate and TS guess fireworks

mjohnson541 committed:
enable automatic postprocessing to be requested

mjohnson541 committed:
check thermo and reaction libraries get made in the workflow test

mjohnson541 committed:
adapt the postprocessing notebook

use the new Configuration and Kinetics postprocessing objects
provide thermochemistry information

mjohnson541 committed:
remove old slab_init

this isn't the slab_init for the slab we are using now so, better to remove on the off chance this causes confusion somewhere

mjohnson541 committed:
enable input of individual species/adsorbates as "molecule"s in the input yaml

this allows addition of individual species without a need to define a reaction that includes them

mjohnson541 committed:
update rxn_test.yaml to improve testing

this both tests the individual species input and adds a gas phase adsorbate

mjohnson541 committed:
fix guess generation for gas phase species

mjohnson541 committed:
enable get_reference_energies to also return what atoms can be referenced from the available reference species

mjohnson541 committed:
fix configuration classes

remove slab from GasConfiguration and give GasConfiguration create_RMG_header
fix create_RMG_header for SurfaceConfiguration

mjohnson541 committed:
avoid outputting thermochemistry for species not properly referenced

this divides the completed species into those thermochemistry can be run for and those that cannot and skip thermochemistry for those species.

mjohnson541 committed:
adapt Postprocessing firework to thermochemistry availability tracking

mjohnson541 committed:
add calculate_thermodynamic_references keyword

this keyword causes Pynta to automatically add the four CHON thermodynamic reference species H2, H2O, CH4 and NH3 to the pynta run (if they are not already available in the Adsorbates folder)

mjohnson541 committed:
adapt generate_pair_geometries to handle input sites and site_adjacency's

mjohnson541 committed:
adapt setup_pair_opts_for_rxns to handle input adsorbates and input sites/site_adjacency

mjohnson541 committed:
enable get_coadsorbate_information to pass imag_freq_max through to get_unique_adsorbate_geometries

mjohnson541 committed:
adapt get_configs_for_calculation to operate on multiple coadsorbates

in particular the input and output dictionaries are now {coad1name: old_dict1, coad2name: old_dict2 ...}

mjohnson541 committed:
adapt CalculateConfigurationEnergiesTask to operate on distinct coadsorbates

mjohnson541 committed:
adapt train_covdep_model_firework to handle multiple coadsorbates

mjohnson541 committed:
adapt SelectCalculationsTask to handle multiple coadsorbates

mjohnson541 committed:
adapt CoverageDependence to handle adsorbates and multiple coadsorbates

also enables passing of imag_freq_max to CoverageDependence

mjohnson541 committed:
enable CoverageDependence to launch in execute

mjohnson541 committed:
adapt postprocessing notebook to thermochemistry changes

postprocessing notebook will now not output thermochemistry if the appropriate references aren't available

mjohnson541 committed:
add a full xtb pynta run for testing coverage dependence

this is a full scale execution of the test CH3OX <=> HOCH2X reaction run that actually finds real TSs plus: H2, O=X, andHX

mjohnson541 committed:
add fireworks configurations for covdep test

mjohnson541 committed:
this adds a set of pairs calculations for the covdep run

These could be calculated within the covdep test, but it takes a lot longer than all other parts of the workflow in the test and tests a lot less of the code than the rest of the workflow

testing the pairs calculations would likely be better tested with an independent and fast test of the generation of the pairs input files that does not involve running all the pairs calculations

mjohnson541 committed:
directory setup and directory cleaning for covdep test

mjohnson541 committed:
add the covdep workflow test

this runs covdep at GFN1-xTB level on the CH3OX <=> HOCH2X reaction with co-adsorbates HX and O=X

to speed the run up we only run one iteration and only run one calculation inside that iteration

imag_freq_max is set to infinity because GFN1-xTB gives at least one large imaginary frequency for all of the isolated coadsorbate configurations and otherwise this example crashes because all the coadsorbate configurations are filtered out

mjohnson541 committed:
skip calculation of thermodynamic references in the Pynta workflow test

mjohnson541 committed:
handle missing frequency files in get_best_adsorbate_xyz

mjohnson541 committed:
upgrade get_best_adsorbate_xyz to compare 2D structures to target

mjohnson541 committed:
fix call of get_best_adsorbate_xyz

mjohnson541 committed:
fix/cleanup task.py imports

mjohnson541 committed:
remove tasks import from coveragedependence.py

mjohnson541 committed:
Have git track test, but not launches in IPython

mjohnson541 committed:
error handling for get_best_adsorbate_xyz

mjohnson541 committed:
move get_best_adsorbate_xyz to geometricanalysis.py

mjohnson541 committed:
move mol_to_atoms to geometricanalysis.py

mjohnson541 committed:
add get_energy_correction_configuration for computing the covdep energy correction for a given configuration

mjohnson541 committed:
add functions for computing barrier and reaction energy covdep corrections

mjohnson541 committed:
add extract_covdep_data for extracting import information for covdep postprocessing

mjohnson541 committed:
add analyze_covdep_lowest_energy function for examining the lowest energy covdep structures of a configuration

mjohnson541 committed:
add analyze_covdep_sample_data for looking at covdep samples associated with a target configuration

mjohnson541 committed:
add coverage dependence postprocessing notebook

mjohnson541 committed:
automatically apply and enforce reaction indexing

mjohnson541 committed:
postprocessing exception handling

mjohnson541 committed:
fix bug related to Xv generation

when we attempt to convert sites with a vdW bond only into a group RMG gives us an atom with an Xv atomtype
that cannot match other sites in the isomorphism so we convert all of these into X atomtypes

mjohnson541 committed:
get adsorbate mappings directly from their info file

the adsorbates get generated in the Adguess fireworks now

mjohnson541 committed:
use single bond comparisons to avoid issues with generating the map for ts

mjohnson541 committed:
throw error if reactant/product is missing in TSguess

mjohnson541 committed:
properly compute num_surf_sites in TSguess

mjohnson541 committed:
allow to keep vdW bonds if they exist in the starting adsorbate for the TSguess

mjohnson541 committed:
skip structures if we cannot find occupied sites matching the template

mjohnson541 committed:
overhaul bond order determination

this changes the electron distribution algorithm to during iterations to always first search for atoms who only
have one way of distributing bonds to achieve octet and satisfy them, only relying on the prioritization scheme
when there are multiple ways to distribute electrons to achieve octet for all remaining atoms

mjohnson541 committed:
skip configurations that cannot be converted to 2D in HFSP

mjohnson541 committed:
skip 2D configurations that cannot be properly labeled from the template in HFSP
(usually means a reactant/product is very wrong)

mjohnson541 committed:
handle ValueError in TS validation

mjohnson541 committed:
add additional print statements to HFSP logging

mjohnson541 committed:
improve missing reactant or product logging

mjohnson541 committed:
fix flipped boolean in fix_bond_orders

mjohnson541 committed:
handle desorption of vdW species when adding them as a coadsorbate

mjohnson541 committed:
compress the adjacency list identifier

this should help avoid errors related to folder name length

mjohnson541 committed:
prefer HFSP directions that have fewer vdW bonded species

mjohnson541 committed:
increment to python 3.9

mjohnson541 committed:
turn off verbosity for GFN1-xTB

mjohnson541 committed:
use python 3.9 in CI

mjohnson541 committed:
tweak CI workflow

mjohnson541 committed:
handle versioning

this keeps behavior the same for pysidt v1.1.0, there's a good chance it is desirable to have
weigh_node_selection_by_occurrence or at least some sort of hybrid algorithm between doing so and not,
but for now we keep it consistent
